### PR TITLE
Feat/support nested field names

### PR DESCRIPTION
**Pull request by jarvisraymond-uchicago**
Jira Ticket: [HP-1084](https://ctds-planx.atlassian.net/browse/HP-1084)

This updates the Discovery app to support nested field names. This requires that the nested fields names be referenced in gitops.json using the syntax supported by jsonpath, for example: ``` "field": "study_metadata.citation.investigators[*].investigator_last_name",```

To support the array returned from jsonpath, the search has been updated to call each item of the studyPreviewTextArray in ```src/Discovery/DiscoveryListView.tsx```

Please note that jssearch has limitations regarding deeply nested search: https://github.com/bvaughn/js-search/pull/78

#### It is able to use nested values for tags: 
image
image

#### Able to find nested values for Last Names: 
image
image

output

#### Able to find nested values using table search
output

### New Features
- Disocvery page: study details fields, discovery table fields, aggregation fields and search fields now supports using [JSONPath syntax](https://github.com/dchester/jsonpath#jsonpath-syntax) to represent nested fields in metadata

### Dependency updates
- Adds: "jsonpath": "^1.1.1",

[HP-1084]: https://ctds-planx.atlassian.net/browse/HP-1084?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by craigrbarnes** [comment]
A couple of notes:
Your additions to the Study Details do work, however the configuration needs to be adjusted:

in gitops.json,

add:
```
                {
                  "type": "text",
                    "label": "Alternative Study Description",
                    "sourceField": "Minimal Info.alternative_study_description",
                    "default": "Not specified"
                },
```
here
image

Note the actual path is `"Minimal Info.alternative_study_description"`

Restart you local data-portal, go to the discovery page and search for "There are no FDA-approved medications for stimulant use disorders" You should get the study titled: **PTPRD ligands for stimulant and opiate use disorders**
If you bring up the study details: it should have Alternative Text entry (ignore the Project Number2 title) 

image

File: src/Discovery/DiscoveryListView.tsx [hunk 1]
```diff
@@ -165,7 +167,7 @@ const DiscoveryListView: React.FunctionComponent<Props> = (props: Props) => {
                     props.setModalVisible(true);
                   }}
                 >
-                  {renderValue(studyPreviewText)}
+                  {studyPreviewTextArray.forEach((item: string | undefined) => renderValue(item))}
```
Comment (mfshao) [inline src/Discovery/DiscoveryListView.tsx:170]:
```suggestion
                  {studyPreviewTextArray.map((item: string | undefined) => renderValue(item))}
```
I think this is the way to get them rendered

**Review by mfshao — COMMENTED** [review]
most of the places renders ok, table columns looks good and study details page rows looks good

the study preview field is not rendered correctly and I think my suggestion would help

there are two more places also needs some update in order to be able to dealing with the nested fields correctly:
1. for `headerField`: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/DiscoveryDetails.tsx#L223
2. for `search`, specifically where the code is loading study data into jssearch: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/Discovery.tsx#L271-L296. The jssearch library appears to be able to deal with nested fields, but it expects to receive an array of strings instead of a dot-separated string, so we need to make that conversion, see https://github.com/bvaughn/js-search/blob/master/README.md?plain=1#L75

**Comment by jarvisraymond-uchicago** [comment]
> most of the places renders ok, table columns looks good and study details page rows looks good
> 
> the study preview field is not rendered correctly and I think my suggestion would help
> 
> there are two more places also needs some update in order to be able to dealing with the nested fields correctly:
> 
>     1. for `headerField`: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/DiscoveryDetails.tsx#L223
> 
>     2. for `search`, specifically where the code is loading study data into jssearch: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/Discovery.tsx#L271-L296. The jssearch library appears to be able to deal with nested fields, but it expects to receive an array of strings instead of a dot-separated string, so we need to make that conversion, see https://github.com/bvaughn/js-search/blob/master/README.md?plain=1#L75

Thanks @mfshao 

I updated the 

> most of the places renders ok, table columns looks good and study details page rows looks good
> 
> the study preview field is not rendered correctly and I think my suggestion would help
> 
> there are two more places also needs some update in order to be able to dealing with the nested fields correctly:
> 
>     1. for `headerField`: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/DiscoveryDetails.tsx#L223
> 
>     2. for `search`, specifically where the code is loading study data into jssearch: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/Discovery.tsx#L271-L296. The jssearch library appears to be able to deal with nested fields, but it expects to receive an array of strings instead of a dot-separated string, so we need to make that conversion, see https://github.com/bvaughn/js-search/blob/master/README.md?plain=1#L75

Thanks @mfshao !

I have updated my PR as follows per your suggestions.

1.) for ```headerField```: around https://github.com/uc-cdis/data-portal/blob/feat/supportNestedFieldNames/src/Discovery/DiscoveryDetails.tsx#L223

I have added a call to use jsonpath: 
```      
<h3 className='discovery-modal__header-text'>
        {jsonpath.query(props.modalData, `$.${headerField}`)}
</h3>
``` 

2.) For the jssearch library, I found formatting the array and then calling search.addIndex on each field worked, whereas calling search.addIndex on the entire array did not. I added a function to remove the wildcard character for jsonpath and to return a new array split by the ``` . ```. I did this for the studyColumns and the studyPreviewField. Let me know if this also needs to be done for the ```searchableFields``` variable. 

Example: 
```
        const studyColumnFieldsArr = formatSearchIndex(column.field);
          studyColumnFieldsArr.forEach((field)=>{
            search.addIndex(field)
          })
```
and 
```
  const formatSearchIndex = (index: String) => {
    const wildCardStringRegex = new RegExp(/\[\*\]/,"g");
    const indexWithoutWildcards = index.replace(wildCardStringRegex,'');
    const indexArr = indexWithoutWildcards.split('.');
    return indexArr;
  }
```

3.) I merged your suggested change to use Map() instead of ForEach() to return an array and output JSX for the ```studyPreviewTextArray```

File: src/Discovery/Discovery.tsx [hunk 2]
```diff
@@ -277,19 +287,25 @@ const Discovery: React.FunctionComponent<Props> = (props: Props) => {
     // in the table and the study description.
     // ---
     const searchableFields = config.features.search.searchBar.searchableTextFields;
-    if (searchableFields) {
+    if (searchableFields && 0==1) {
```
Comment (mfshao) [inline src/Discovery/Discovery.tsx:290]: we need to ass this back, also the code needs to be able to handle nested fields in these `searchableFields`

File: src/Discovery/Discovery.tsx [hunk 3]
```diff
@@ -279,17 +288,20 @@ const Discovery: React.FunctionComponent<Props> = (props: Props) => {
     const searchableFields = config.features.search.searchBar.searchableTextFields;
     if (searchableFields) {
       searchableFields.forEach((field) => {
-        search.addIndex(field);
+        const formattedFields = formatSearchIndex(field);
+        formattedFields.forEach((formattedField) => search.addIndex(formattedField));
```
Comment (mfshao) [inline src/Discovery/Discovery.tsx:292]:
```suggestion
        search.addIndex(formattedFields);
```
During my testing, this works, but the way of adding each individual field into `search` doesn't

File: src/Discovery/Discovery.tsx [hunk 4]
```diff
@@ -279,17 +288,20 @@ const Discovery: React.FunctionComponent<Props> = (props: Props) => {
...
       });
     } else {
       config.studyColumns.forEach((column) => {
         if (!column.contentType || column.contentType === 'string') {
-          search.addIndex(column.field);
+          const studyColumnFieldsArr = formatSearchIndex(column.field);
+          studyColumnFieldsArr.forEach((studyColumnField) => search.addIndex(studyColumnField));
```
Comment (mfshao) [inline src/Discovery/Discovery.tsx:298]:
```suggestion
          search.addIndex(studyColumnFieldsArr);
```
same as above

File: src/Discovery/Discovery.tsx [hunk 5]
```diff
@@ -279,17 +288,20 @@ const Discovery: React.FunctionComponent<Props> = (props: Props) => {
...
         }
       });
       // Also enable search over preview field if present
       if (config.studyPreviewField) {
-        search.addIndex(config.studyPreviewField.field);
+        const studyPreviewFieldArr = formatSearchIndex(config.studyPreviewField.field);
+        studyPreviewFieldArr.forEach((studyPreviewField) => search.addIndex(studyPreviewField));
```
Comment (mfshao) [inline src/Discovery/Discovery.tsx:304]:
```suggestion
        search.addIndex(studyPreviewFieldArr);
```
same as above

File: src/Discovery/Discovery.tsx (hunk 3 above)
Comment (jarvisraymond-uchicago) [reply src/Discovery/Discovery.tsx]: Suggested applied! Thank you.

File: src/Discovery/Discovery.tsx (hunk 4 above)
Comment (jarvisraymond-uchicago) [reply src/Discovery/Discovery.tsx]: Suggestion applied! Thank you.

File: src/Discovery/Discovery.tsx (hunk 5 above)
Comment (jarvisraymond-uchicago) [reply src/Discovery/Discovery.tsx]: Suggestion applied! Thank you.

**Comment by mfshao** [comment]
search works, but "study description summary" didn't appear in qa-heal's study details page

**Comment by mfshao** [comment]
I think you need also update the logic in this line https://github.com/uc-cdis/data-portal/pull/1324/files#diff-353c27bcc37c1ad1a11d316dcb4511a986dc1c449400b47fd0dd50354f6ddd1eR181 so it can uses jsonpath to determine whether a section has values or not

**Comment by mfshao** [comment]
> 7f440bf

addressed by [7f440bf](https://github.com/uc-cdis/data-portal/pull/1324/commits/7f440bfe412bd4f69b522df0edb84884a87542a9)

**Comment by mfshao** [comment]
> I think you need also update the logic in this line https://github.com/uc-cdis/data-portal/pull/1324/files#diff-353c27bcc37c1ad1a11d316dcb4511a986dc1c449400b47fd0dd50354f6ddd1eR181 so it can uses jsonpath to determine whether a section has values or not

addressed by [7f440bf](https://github.com/uc-cdis/data-portal/pull/1324/commits/7f440bfe412bd4f69b522df0edb84884a87542a9)

**Review by mfshao — APPROVED** [review]

**Review by mfshao — APPROVED** [review]

**Pull request merged**